### PR TITLE
SWATCH-2069: Fix IncorrectResultSizeDataAccessException when saving subs

### PR DESCRIPTION
**Pull request by Sgitario**
Jira issue: [SWATCH-2069](https://issues.redhat.com/browse/SWATCH-2069)

## Description
The methods "findBySubscriptionNumber" and "findActiveSubscription" expect to find unique results. When multiple entities are found, it returns the IncorrectResultSizeDataAccessException exception.

## Testing

1.- podman-compose up
2.- apply migrations

```
./gradlew :liquibaseUpdate
```

3.- add the required data:

which is the offering and the duplicate subscription with a different start date. 
```
INSERT INTO OFFERING(sku,product_name,product_family,sla,description,has_unlimited_usage) 
VALUES ('MW00332','OpenShift Data Science','OPENSHIFT ENTERPRISE','Premium','Red Hat OpenShift Data Science','false');

INSERT INTO subscription(sku,org_id,subscription_id,quantity,start_date,subscription_number)
VALUES('MW00332','16764251','12396617','25','2023-02-02T02:13:35.983189Z','12396774');

INSERT INTO subscription(sku,org_id,subscription_id,quantity,start_date,subscription_number)
VALUES('MW00332','16764251','12396617','25','2023-02-06T02:13:35.983189Z','12396774');
```

4.- (only for local testing) start ActiveMQ:

```
docker run --detach --name mycontainer -p 61616:61616 -p 8161:8161 --rm apache/activemq-artemis:latest-alpine
```

5.- start the service:

```
DEV_MODE=true SUBSCRIPTION_USE_STUB=true USER_USE_STUB=true SPRING_PROFILES_ACTIVE="api,worker,kafka-queue,capacity-ingress" PRODUCT_USE_STUB=true UMB_ENABLED=true SPRING_ACTIVEMQ_BROKER_URL=tcp://localhost:61616 SPRING_ACTIVEMQ_USER=artemis SPRING_ACTIVEMQ_PASSWORD=artemis ./gradlew :bootRun
```

6.- Send the payload via artemis management site: http://localhost:8161/console/artemis/artemisSendMessage?nid=root-org.apache.activemq.artemis-0.0.0.0-addresses-Consumer.null.swatch-null-VirtualTopic_canonical_subscription.VirtualTopic.canonical.subscription-queues-anycast-Consumer.null.swatch-null-VirtualTopic_canonical_subscription.VirtualTopic.canonical.subscription

Payload:

```xml
<CanonicalMessage xmlns="http://esb.redhat.com/Canonical/6">
    <Header>
        <System>SUBSCRIPTION</System>
        <Operation>update</Operation>
        <Type>Subscription</Type>
        <InstanceId>6446c7c050d3b36402663da3</InstanceId>
        <Timestamp>2023-04-24T14:17:41.307</Timestamp>
    </Header>
    <Payload>
        <Sync>
            <Subscription>
                <CreatedBy>OSUBPRODTEST3EAC72C765804E2B8D34484E7FF8D1C8</CreatedBy>
                <CreatedDate>2023-04-24T14:11:16.809</CreatedDate>
                <LastUpdatedBy>RHMIDDLEWARE</LastUpdatedBy>
                <LastUpdatedDate>2023-04-24T14:15:18.000</LastUpdatedDate>
                <Identifiers>
                    <MasterSystem>SUBSCRIPTION</MasterSystem>
                    <Identifier entity-name="Install Base"
                        qualifier="number" system="EBS">39832705</Identifier>
                    <Identifier entity-name="Registration" qualifier="number">e401bce70edd43bd</Identifier>
                    <Identifier entity-name="Subscription"
                        qualifier="number" system="SUBSCRIPTION">12396774</Identifier>
                    <Reference entity-name="Installation" qualifier="number">c967ef3956ff3def</Reference>
                    <Reference entity-name="Customer" qualifier="number" system="EBS">customer123</Reference>
                    <Reference entity-name="Account" qualifier="number" system="EBS">account123</Reference>
                    <Reference entity-name="Customer" qualifier="id" system="WEB">16764251</Reference>
                    <Reference entity-name="Account" qualifier="id" system="WEB">16764251</Reference>
                </Identifiers>
                <Status>
                    <State>Active</State>
                    <StartDate>2023-04-24T00:00:00.000</StartDate>
                </Status>
                <Quantity>1</Quantity>
                <effectiveStartDate>2023-04-24T00:00:00.000</effectiveStartDate>
                <effectiveEndDate>2026-04-23T00:00:00.000</effectiveEndDate>
                <Product serviceable="false">
                    <CreatedBy>OSUBPRODTEST3EAC72C765804E2B8D34484E7FF8D1C8</CreatedBy>
                    <CreatedDate>2023-04-24T14:11:16.809</CreatedDate>
                    <Identifiers>
                        <MasterSystem>EBS</MasterSystem>
                        <Identifier entity-name="Product"
                            qualifier="number" system="SUBSCRIPTION">37188549</Identifier>
                    </Identifiers>
                    <Sku>BASILISK</Sku>
                    <InventoryOperatingUnit>
                        <Number>118</Number>
                    </InventoryOperatingUnit>
                    <Product serviceable="true">
                        <CreatedBy>OSUBPRODTEST3EAC72C765804E2B8D34484E7FF8D1C8</CreatedBy>
                        <CreatedDate>2023-04-24T14:11:16.809</CreatedDate>
                        <LastUpdatedBy>SYSADMIN</LastUpdatedBy>
                        <LastUpdatedDate>2023-04-24T14:17:02.000</LastUpdatedDate>
                        <Identifiers>
                            <MasterSystem>EBS</MasterSystem>
                            <AuthoringOperatingUnit>
                                <Number>103</Number>
                            </AuthoringOperatingUnit>
                            <Identifier entity-name="Product"
                                qualifier="number" system="SUBSCRIPTION">37188550</Identifier>
                            <Reference entity-name="Contract"
                                qualifier="number" system="EBS">16612423</Reference>
                            <Reference entity-name="Contract"
                                qualifier="id" system="EBS">25565753</Reference>
                        </Identifiers>
                        <Status>
                            <State>Active</State>
                            <StartDate>2023-04-24T00:00:00.000</StartDate>
                            <EndDate>2026-04-23T00:00:00.000</EndDate>
                        </Status>
                        <Status>
                            <State>Signed</State>
                            <StartDate>2023-04-24T00:00:00.000</StartDate>
                        </Status>
                        <Status primary="true">
                            <State>Terminated</State>
                            <StartDate>2023-04-24T14:17:02.000</StartDate>
                        </Status>
                        <Sku>SVC2681</Sku>
                        <InventoryOperatingUnit>
                            <Number>118</Number>
                        </InventoryOperatingUnit>
                        <ContractDescription>Reg Number activation from Hock</ContractDescription>
                        <ContractHeaderStatus>Terminated</ContractHeaderStatus>
                    </Product>
                </Product>
            </Subscription>
        </Sync>
    </Payload>
</CanonicalMessage>
```

7.- Verification:

You should see the following trace in the service logs:

```Received UMB message... ```

And no exceptions.

Before these changes, we were seeing the IncorrectResultSizeDataAccessException exceptions.

### CODE REVIEW

File: src/main/java/org/candlepin/subscriptions/subscription/ActiveMQServiceConfiguration.java [hunk 1]
```diff
@@ -51,6 +51,14 @@ public ActiveMQSslConnectionFactory activeMQSslConnectionFactory(
       ActiveMQProperties activeMQProperties) throws Exception {
     ActiveMQSslConnectionFactory factory = new ActiveMQSslConnectionFactory();
     factory.setExceptionListener(e -> log.error("Exception thrown in ActiveMQ connection", e));
+    if (StringUtils.hasText(activeMQProperties.getUser())) {
```
Comment (Sgitario) [inline src/main/java/org/candlepin/subscriptions/subscription/ActiveMQServiceConfiguration.java:54]: I had to include these changes, so I can verify the changes using a remote ActiveMQ connection (from a container).

**Comment by lindseyburnett** [comment]
The only question I have with this approach is if we're going to get any undesirable behavior by changing the db query to "give me the most recent subscription record by subscription number".  I think it's probably fine, but any idea what the differences in data are with the multiple results?  pinging @kahowell for thoughts

**Comment by Sgitario** [comment]
> The only question I have with this approach is if we're going to get any undesirable behavior by changing the db query to "give me the most recent subscription record by subscription number". I think it's probably fine, because I suppose saving something is better than the exceptions we're getting...but any idea what the differences in data are with the multiple results? pinging @kahowell for thoughts

I attached in slack a csv with the differences in data for a single subscription (
[sub_12396774.csv](https://github.com/RedHatInsights/rhsm-subscriptions/files/13897967/sub_12396774.csv)), but yes, we need a better analysis if the duplicates are expected or a side effect of a wrong logic somewhere. My proposal here was to quick fix the exception and then in a follow-up task investigate the data issue better.

**Comment by lindseyburnett** [comment]
> > The only question I have with this approach is if we're going to get any undesirable behavior by changing the db query to "give me the most recent subscription record by subscription number". I think it's probably fine, because I suppose saving something is better than the exceptions we're getting...but any idea what the differences in data are with the multiple results? pinging @kahowell for thoughts
> 
> I attached in slack a csv with the differences in data for a single subscription ( [sub_12396774.csv](https://github.com/RedHatInsights/rhsm-subscriptions/files/13897967/sub_12396774.csv)), but yes, we need a better analysis if the duplicates are expected or a side effect of a wrong logic somewhere. My proposal here was to quick fix the exception and then in a follow-up task investigate the data issue better.

works for me, can you send me the card number for that follow up task once you've got it? thanks!

**Review by lindseyburnett — APPROVED** [review]

**Comment by Aurobinda55** [comment]
@Sgitario I am not able to reproduce the issue on EE as UMB is not enabled on EE and we don’t have an internal endpoint to send a custom UMB message. You can merge the PR and I will check the error logs if exist in stage and prod and update once the changes are in stage and prod.

**Review by Aurobinda55 — APPROVED** [review]

**Pull request merged**